### PR TITLE
fix(pet-35): NFKC + homoglyph + casefold in tool name normalization

### PR DESCRIPTION
**Pull request by ziomancer**
## Summary
- Harden `ToolCallGuard._normalize_tool_name()` with NFKC normalization, homoglyph mapping, and `casefold()` to close a defense-in-depth gap where Cyrillic/fullwidth confusables could bypass alias resolution
- Reorder operations so `strip()` runs before alias lookup — whitespace-padded tool names now correctly resolve aliases
- Harmonize `profiles/__init__.py` `.lower()` → `.casefold()` for consistency between guard runtime and profile construction

## Layered defense

The normalization pipeline applies six steps in order: strip → NFKC → homoglyph → casefold → namespace strip → alias lookup. NFKC collapses fullwidth chars, the homoglyph table maps Cyrillic/Greek confusables to Latin, and casefold handles Unicode-correct case folding. The GUARD-03 alias→exempt defense is preserved with casefold semantics. The return value is always `resolved.strip().casefold()` ensuring downstream consumers see fully normalized tool names.

## Files changed

| File | Change |
|------|--------|
| `petasos/premium/guard.py` | Rewrite `_normalize_tool_name`: new operation order, import `_HOMOGLYPH_TABLE` + `unicodedata`, GUARD-03 preserved with casefold |
| `petasos/premium/profiles/__init__.py` | 4 `.lower()` → `.casefold()` swaps for exempt-list and collision-check consistency |
| `tests/adversarial/guard/test_tool_smuggling.py` | 4 new smuggling tests + 1 existing test updated |
| `tests/test_guard.py` | 5 new normalization tests + 1 existing test updated |

## Test plan
- [x] `C:\python310\python.exe -m pytest tests/test_guard.py tests/adversarial/guard/test_tool_smuggling.py -v` — 56/56 pass (full output: docs/specs/TODO/PET-35.test-output.txt)
- [x] `ruff check .` — clean
- [x] `mypy --strict .` — clean
- [x] Full suite: 655 passed, 28 skipped, 2 pre-existing benchmark errors

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved tool-name normalization: consistent Unicode handling (NFKC, homoglyphs), case-folding, and correct ordering of whitespace and namespace stripping; predictable empty/whitespace results and alias collision checks.

* **Tests**
  * Expanded coverage for normalization edge cases, including homoglyphs, mixed scripts, namespace variants, empty/whitespace inputs, and regression checks.

* **Documentation**
  * Added captured test-run output documenting a successful test suite run.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Vigil-Harbor/Petasos/pull/27?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: docs/specs/TODO/PET-35.test-output.txt [hunk 1]
```diff
@@ -0,0 +1,67 @@
+﻿============================= test session starts =============================
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply docs/specs/TODO/PET-35.test-output.txt]: Resolved in c0c403b

File: petasos/premium/guard.py [hunk 2]
```diff
@@ -170,18 +172,15 @@ def _normalize_tool_name(self, tool_name: str) -> str:
             resolved != pre_alias
             and self._profile
             and name in self._profile.tool_alias_map
-            and resolved.strip().lower() in self._profile.tool_exempt_list
+            and resolved.strip().casefold() in self._profile.tool_exempt_list
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply petasos/premium/guard.py:177]: Skipped — profile alias values are operator-controlled data, not attacker input. The spec (Decision 7) explicitly deferred full NFKC+homoglyph normalization on profile data as a pre-existing gap that PET-35 does not worsen. The return path already applies strip().casefold() (Decision 8). A follow-up ticket can add full pipeline normalization to profile parsing if non-ASCII profile definitions become a supported use case.

File: petasos/premium/profiles/__init__.py [hunk 3]
```diff
@@ -76,10 +76,10 @@ def _parse_profile(data: dict[str, Any]) -> ResolvedProfile:
             raise ValueError("tool_alias_map values must be non-empty strings")
     alias_map = {k: v.strip() for k, v in alias_map.items()}
 
-    exempt_set = frozenset(s.strip().lower() for s in data.get("tool_exempt_list", []))
+    exempt_set = frozenset(s.strip().casefold() for s in data.get("tool_exempt_list", []))
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply petasos/premium/profiles/__init__.py:79]: Skipped — exempt list entries come from profile JSON definitions (operator-controlled, not user input). All current profiles use ASCII values. The spec explicitly scoped profiles changes to .lower()→.casefold() harmonization only (Decision 6). Applying NFKC+homoglyph to profile data at parse time would require importing the homoglyph table into profiles and is deferred per Decision 7.

File: petasos/premium/profiles/__init__.py [hunk 4]
```diff
@@ -76,10 +76,10 @@ def _parse_profile(data: dict[str, Any]) -> ResolvedProfile:
...
 
     # GUARD-03: a profile alias may not target one of its own exempt keys
-    collisions = {v.lower() for v in alias_map.values()} & exempt_set
+    collisions = {v.casefold() for v in alias_map.values()} & exempt_set
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply petasos/premium/profiles/__init__.py:82]: Skipped — same root cause as the exempt-list finding. The bypass scenario requires an operator to define a Cyrillic alias target in profile JSON, which is operator error, not an attack vector. Pre-existing gap deferred per spec Decision 7.

File: tests/adversarial/guard/test_tool_smuggling.py [hunk 5]
```diff
@@ -21,16 +21,59 @@ def _guard_with_profile(profile: ResolvedProfile) -> ToolCallGuard:
...
+
+def test_cyrillic_a_in_bash_normalizes() -> None:
+    """GUARD-02: Cyrillic а (U+0430) in 'bаsh' maps to 'exec' via homoglyph table."""
+    from petasos.config import PetasosConfig
+    from petasos.pipeline import Pipeline
+    from petasos.premium.frequency import FrequencyTracker
+
+    cfg = PetasosConfig()
+    guard = ToolCallGuard(Pipeline(config=cfg), FrequencyTracker(cfg), cfg)
+    assert guard._normalize_tool_name("bаsh") == "exec"
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply tests/adversarial/guard/test_tool_smuggling.py:43]: Skipped — the test passes (56/56). _HOMOGLYPH_TABLE is built via str.maketrans() in petasos/normalize.py (line 48), which returns an ordinal-keyed mapping. str.translate() works correctly with it. The search script looked for the table definition in guard.py rather than normalize.py where it is defined and imported from.
